storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -4017,122 +4014,1889 @@ dberr_t fil_space_t::garbage_collect(bool shutdown)
+  }
+};
 
-  /* Set fixed size as threshold to truncate */
-  if (fixed_size > last_used_extent)
-    last_used_extent= fixed_size;
+/** page operation for the system tablespace does the 2 things:
+1) Page Allocation
+2) Page removal
 
-  bool old_dblwr_buf= buf_dblwr.in_use();
-  /* Flush all pages in buffer pool, so that it doesn't have to
-  use doublewrite buffer and disable dblwr and there should
-  be enough space in redo log */
-  log_make_checkpoint();
-  fil_system.set_use_doublewrite(false);
+Steps for page allocation depends on new extent state.
 
-  buf_block_t *header= nullptr;
-  ut_ad(!fsp_tablespace_validate(space));
+(1) If the xdes_get_state(new_descr) == XDES_FREE then
+remove the new extent from FSP_FREE list
 
-  mtr.start();
-  mtr.x_lock_space(space);
+    (1.1) If the page has to be allocated for segment then
+          add the newly allocated extent descriptor to
+          FSEG_NOT_FULL list and make the xdes_set_state(new_descr)
+          as XDES_FSEG
+
+    (1.2) If the page has to be non-segment page then add the
+          newly allocated extent descriptor to FSP_FREE_FRAG list
+          and make the xdes_set_state(new_descr) as XDES_FREE_FRAG
+
+    (1.3) return the free page from the new extent
+
+(2) If the xdes_get_state(new_descr) == XDES_FREE_FRAG then
+
+     (2.1) Allocate a page from the new extent
+
+     (2.2) xdes_get_n_used(new_descr) is FSP_EXTENT_SIZE then
+           - Remove the new extent descriptor from FSP_FREE_FRAG list
+           - Add the new extent descriptor to FSP_FULL_FRAG list
+             and make xdes_set_state(new_descr) as XDES_FULL_FRAG
+
+(3) If the xdes_get_state(new_descr) == XDES_FSEG then
 
+     (3.1) Allocate a page from extent
+
+     (3.2) xdes_get_n_used(new_descr) is FSP_EXTENT_SIZE then
+            - Remove the new extent descriptor from FSEG_NOT_FULL list
+            - Add the new extent descriptor to FSEG_FULL list
+
+
+Steps for removing the page from extent:
+
+  (1) If the xdes_get_state(m_old_descr) is XDES_FSEG then
+     (1.1) If the number of used pages in descriptor is
+          FSP_EXTENT_SIZE then move the extent descriptor
+          from FSEG_FULL to FSEG_NOT_FULL
+     (1.2) If the number of used page in descriptor to 0 then
+           move the extent descriptor from FSEG_NOT_FULL to FSP_FREE
+     (1.3) Free the page and mark the XDES_FREE_BIT of the
+           respective page in current extent descriptor
+
+  (2) If the xdes_get_stats(m_old_descr) is
+      XDES_FREE_FRAG / XDES_FULL_FRAG then
+
+     (2.1) If the number of used pages in extent descriptor
+           is FSP_EXTENT_SIZE then move the extent descriptor
+           from FSP_FULL_FRAG to FSP_FREE_FRAG list
+     (2.2) If the number of used pages in extent descriptor is 0 then
+           move the extent descriptor from FSP_FREE_FRAG to FSP_FREE
+     (2.3) Free the page and mark the XDES_FREE_BIT of the respective
+           page in current extent descriptor
```

Comment:
These steps are essentially the same as (1.2) and (1.3), only (1.2) and (2.2) are differing between `FSEG_NOT_FULL`  and `FSEG_FREE_FRAG`. That term would be mentioned the previous step. We could simply say "move the extent descriptor to FSP_FREE_FRAG" and avoid mentioning where it is being moved from.

Can you rephrase this comment so that there are alternatives (1a) and (1b), but the (2) and (3) would be the same for both?